lua/ttt2/libraries/game_effects.lua

```diff
@@ -0,0 +1,342 @@
+	return (ent.GetClass and table.HasValue(gameEffects.extinguishables, ent:GetClass()))
+		or ent:IsOnFire()
+end
+
+---
+-- A hook provided by Rubat's Fire Extinguisher, which many other addons implement or hook into.
+-- @param Entity ent The entity to attempt extinguishing.
+-- @return boolean|nil Return true to prevent extinguishing.
+-- @ref https://steamcommunity.com/sharedfiles/filedetails/?id=104607228
+-- @realm shared
+function GM:ExtinguisherDoExtinguish(ent)
+
+end
+
+---
+-- Take an entity that is on fire, and make it not on fire.
+-- @param Entity ent The entity to remove flames from.
+-- @return boolean If the entity was extinguished.
+-- @realm shared
+function gameEffects.Extinguish(ent)
+	local class = ent:GetClass()
+
+	---
+	-- @realm shared
+	if hook.Run("ExtinguisherDoExtinguish", ent) == true then return false end
+
+	local didExtinguish = false
+	if SERVER then
+		if ent:IsOnFire() or ent:IsFlagSet( FL_ONFIRE ) then
+			ent:Extinguish()
+			ent:RemoveFlags( FL_ONFIRE )
+			didExtinguish = true
+		end
+
+		if string.find(class, "env_fire") then
+			ent:Fire("Extinguish")
+			didExtinguish = true
+		elseif string.find(class, "ent_minecraft_torch") and ent:GetWorking() then
+			ent:SetWorking(false)
+			didExtinguish = true
+		end
+	end
+
+	if SERVER and gameEffects.IsExtinguishableEntity(ent) then
+		ent:Remove()
+		didExtinguish = true
+	end
+
+	if didExtinguish then
+		sound.Play(Sound("ttt2/sfx/extinguish_fizzle.wav"), ent:GetPos(), 65, 144, 1)
+	end
+
+	return didExtinguish
+end
+
+---
+-- @param number pos The location to attempt extinguishing around.
+-- @param number radius The radius to extinguish all entities found within.
+-- @realm shared
+function gameEffects.ExtinguishInRadius(pos, radius)
+	local foundEnts = ents.FindInSphere(pos, radius)
+	for i = 1, #foundEnts do
+		gameEffects.Extinguish(foundEnts[i])
+	end
+end
+
+---
+-- @param Vector pos The point in question.
+-- @return boolean Whether the provided point was inside the radius of an active smoke.
+-- @realm shared
+function gameEffects.IsPointInSmoke(pos)
+	local smokes = ents.FindByClass("ttt_smoke")
+	for i = 1, #smokes do
+		local smoke = smokes[i]
+		if smoke:GetAlive() and pos:DistToSqr(smoke:GetPos()) < (smoke:GetRadius() ^ 2) then
+			return true
+		end
+	end
+	return false
+end
+
+---
+-- @param table data
+-- @note Structure of data = {
+-- 	position --[[Vector]],
+-- 	shared_seed --[[string]],
+-- 	life_time --[[number]],
+-- 	fade_time --[[number]],
+-- 	random_color --[[boolean]],
+-- 	extinguish_fires --[[boolean]],
+-- 	density --[[number]],
+-- 	radius --[[number]],
+-- 	color_base --[[number]],
+-- 	color_variance --[[number]],
+-- 	start_size_base --[[number]],
+-- 	start_size_variance --[[number]],
+-- 	end_size_base --[[number]],
+-- 	end_size_variance --[[number]],
+--  set_color --[[nil|Color]], -- if set, overrides final color, if omitted, smoke generation is ran and passed in
+-- }
+-- @todo externalize this DTO somewhere better?
```

Comment:
on a second thought: Why a data table here and param lists for anything else in this library? i don't really have a preference for either way but i think we should make it consistent

Reply:
I used a table here and ideally also for some of the fire ones because it's an overwhelming amount of arguments to list in an ordinary function call, and some of them can be given fall-through logic but not all, and although they're _relatively_ in an order of importance it's hard for me to say exactly.

I would split it between the "essentials" as parameters and "optional" values all as the lump data table after having a second look at it, but that might make client code more unwieldy. I'm not sure.